snapcraft/plugins/v2/colcon.py

```diff
@@ -0,0 +1,247 @@
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""The colcon plugin for ROS2 parts.
+
+    - colcon-packages:
+      (list of strings)
+      List of colcon packages to build. If not specified, all packages in the
+      workspace will be built. If set to an empty list ([]), no packages will
+      be built, which could be useful if you only want ROS debs in the snap.
+
+    - colcon-packages-ignore:
+      (list of strings)
+      List of colcon packages to ignore. If not specified or set to an empty
+      list ([]), no packages will be ignored.
+
+    - colcon-cmake-args:
+      (list of strings)
+      Arguments to pass to cmake projects. Note that any arguments here which match
+      colcon arguments need to be prefixed with a space. This can be done by quoting
+      each argument with a leading space.
+
+    - colcon-catkin-cmake-args:
+      (list of strings)
+      Arguments to pass to catkin packages. Note that any arguments here which match
+      colcon arguments need to be prefixed with a space. This can be done by quoting
+      each argument with a leading space.
+
+    - colcon-ament-cmake-args:
+      (list of strings)
+      Arguments to pass to ament_cmake packages. Note that any arguments here which
+      match colcon arguments need to be prefixed with a space. This can be done by
+      quoting each argument with a leading space.
+
+    This plugin expects the build-environment `ROS_DISTRO` and `ROS_BUILD_BASE`
+    to be populated by the `ros2-<distro>` extension.
+
+    This plugin also expects certain variables that are specified by the extension,
+    specific to the ROS distro.  If not using the extension, set these in your
+    `build-environment`:
+      - ROS_DISTRO: "foxy"
+"""
+
+import os
+import pathlib
+import subprocess
+import sys
+from typing import Any, Dict, List, Set
+
+import click
+from catkin_pkg import packages as catkin_packages
+
+from snapcraft.internal.repo import Ubuntu
+from snapcraft.plugins.v1._ros.rosdep import _parse_rosdep_resolve_dependencies
+from snapcraft.plugins.v2 import PluginV2
+
+
+class ColconPlugin(PluginV2):
+    @classmethod
+    def get_schema(cls) -> Dict[str, Any]:
+        return {
+            "$schema": "http://json-schema.org/draft-04/schema#",
+            "type": "object",
+            "additionalProperties": False,
+            "properties": {
+                "colcon-ament-cmake-args": {
+                    "type": "array",
+                    "minitems": 1,
```

Comment:
IIRC, long standing bug in other places, but if the statement is not minItems then it is completely ignored

Reply:
notice the camel casing I

Reply:
hah, well i suppose minItems should be 0 then :D did that too